packages/vite/src/node/utils.ts

```diff
@@ -1267,3 +1267,8 @@ export function arrayEqual(a: any[], b: any[]): boolean {
   }
   return true
 }
+
+export function evalValue<T = any>(rawValue: string): T {
+  const fn = new Function(`return (\n${rawValue}\n)`)
+  return fn()
```

Comment:
Nice trick! My only worry is if this could be an attack vector. The user could add any function call and it would be evaluated, no? Do we have other places where we do the same?

Reply:
I don't think this would be an attack vector though since it's extracting the code you're writing. If it comes from an npm package, the risk is the same as installing a malicious package, e.g. the attacker could run malicious code directly `something()` instead of crafting a `import.meta.glob(..., { run: something() })` contraption.

We currently also use `new Function` for SSR transform.

Reply:
> We currently also use `new Function` for SSR transform.

I don't see a similar use case where the user can feed any input. Do you have a link?

But anyways, I think you are right, the project could directly have the malicious code in a plugin. This isn't something new.

Any reason you are keeping it as a draft? We could merge and let people test it out if you think it is ready.

Reply:
> I don't see a similar use case where the user can feed any input. Do you have a link?

It's `new AsyncFunction` 😅  at [here](https://github.com/vitejs/vite/blob/7b7206951fb93ea7e2555c0bb152be8c85220f30/packages/vite/src/node/ssr/ssrModuleLoader.ts#L193-L201).

> Any reason you are keeping it as a draft? We could merge and let people test it out if you think it is ready.

I think I was just unsure about this approach, and probably the only downside I see so far is that `acorn` can give finer grain error messages when things are non a literal (currently it doesn't do that though but it has the potential).

I'll clean this up then and mark for review.

Reply:
My only concern is that having access to (Vite's process's) global variables could lead to people understanding them incorrectly.
(e.g. `import.meta.glob('./foo', { as: process.versions.node })`, `import.meta.glob('./foo', { as: (() => { process.env.FOO = 'foo'; return 'url' })() })`)

I think this will do the job.
```js
new Function(`
  var console, exports, global, module, process, require
  return (() => (\n${rawValue}\n)()
`)
```
We could use `node:vm` to make it more accurate but I don't we need it.

Also I think it's possible to change prototype but I assume nobody we do that. (e.g. `import.meta.glob('./foo', { as: (() => { String.prototype.toLowerCase = () => 'aa'; return 'url'.toLowerCase() } })`)

Reply:
Adding the globals sounds like a good safeguard. I think as long as we declare that we only support static options, IIFE could be considered not supported and "use at your own risk".

I'm not sure why we need to do `return (() => (\n${rawValue}\n)()` though, we're already inside a function?

Reply:
> I'm not sure why we need to do `return (() => (\n${rawValue}\n)()` though, we're already inside a function?

Ah, yes, we don't need to wrap.
I added them to avoid the syntax error (`SyntaxError: Identifier 'console' has already been declared`).
```js
var console
let console
```
But yes, given that we're wrapping with `()` and cannot declare a variable in the same scope, we don't need that 👍